### PR TITLE
Add a test to validate the problem with first_found

### LINKED ISSUE #70772: with_first_found fails to find a file if an undefined variable is used in the list since ansible 2.9
##### SUMMARY

with_first_found fails "No file was found when using first_found. ..." even if a file matches search conditions, when an undefined variable is used in files parameter.
This is started after I've upgraded from ansible 2.8 to 2.9, so I suppose it is a bug.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
first_found

##### ANSIBLE VERSION
```paste below
ansible 2.9.10
  config file = None
  configured module search path = ['/Users/afunix/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.9.10/libexec/lib/python3.8/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.8.4 (default, Jul 14 2020, 02:58:48) [Clang 11.0.3 (clang-1103.0.32.62)]
```

##### CONFIGURATION

```paste below
DEFAULT_VAULT_PASSWORD_FILE(env: ANSIBLE_VAULT_PASSWORD_FILE) = /Users/afunix/.vault
```

##### OS / ENVIRONMENT
Host: macOS 10.15.5
Target: CentOS 7.8

##### STEPS TO REPRODUCE

playbook.yml
```yaml
---
- hosts: all
  roles:
      - test

```

roles/test/tasks/main.yml
```yaml
---
- debug:
      msg: "path: {{role_path}}/vars"

- debug:
    msg: "file: {{ansible_distribution}}-{{ansible_distribution_major_version}}.yml"

- name: Load variables specific for OS family
  include_vars: "{{item}}"
  with_first_found:
    - files:
        - "{{ansible_lsb.id}}-{{ansible_lsb.major_release}}.yml"
        - "{{ansible_distribution}}-{{ansible_distribution_major_version}}.yml"
      paths:
        - "{{role_path}}/vars"

```

roles/test/vars/CentOS-7.yml
```yaml
---
test: test
```

##### EXPECTED RESULTS
ansible-playbook is able to find CentOS-7.yml when running against CentOS 7.8 system, since the file exists.

##### ACTUAL RESULTS
ansible-playbook cannot find CentOS-7.yml when running against CentOS 7.8 system.

The issue goes away once I remove "{{ansible_lsb.id}}-{{ansible_lsb.major_release}}.yml" from files.

Checked ansible -m setup for the host and ansible_lsb is empty.

```paste below
$ ansible-playbook -i inventory playbook.yml 

PLAY [all] *****************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *****************************************************************************************************************************************************************************************
ok: [alice]

TASK [test : debug] ********************************************************************************************************************************************************************************************
ok: [alice] => {
    "msg": "path: /Users/afunix/tmp/ansible-with-first-found/roles/test/vars"
}

TASK [test : debug] ********************************************************************************************************************************************************************************************
ok: [alice] => {
    "msg": "file: CentOS-7.yml"
}

TASK [test : Load variables specific for OS family] ************************************************************************************************************************************************************
fatal: [alice]: FAILED! => {"msg": "No file was found when using first_found. Use errors='ignore' to allow this task to be skipped if no files are found"}

PLAY RECAP *****************************************************************************************************************************************************************************************************
alice                      : ok=3    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

$ stat /Users/afunix/tmp/ansible-with-first-found/roles/test/vars/CentOS-7.yml
16777220 38020305 -rw-r--r-- 1 afunix staff 0 15 "Jul 20 18:11:55 2020" "Jul 20 18:11:38 2020" "Jul 20 18:11:38 2020" "Jul 20 18:11:38 2020" 4096 8 0 /Users/afunix/tmp/ansible-with-first-found/roles/test/vars/CentOS-7.yml

$ ansible -i inventories/production/hosts -m setup alice | grep ansible_lsb
        "ansible_lsb": {},
```

### PR DESCRIPTION
**Pull request by bmillemathias**
##### SUMMARY
Fixes #70772 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
first_found

##### ADDITIONAL INFORMATION
code from @sivel

### CODE REVIEW

File: lib/ansible/executor/task_executor.py [hunk 1]
```diff
@@ -224,7 +225,15 @@ def _get_loop_items(self):
...
                                                          convert_bare=False)
                 if not fail:
-                    loop_terms = [t for t in loop_terms if not templar.is_template(t)]
+                    intermediate = []
+                    for term in loop_terms:
+                        if isinstance(term, Mapping) and 'files' in term:
+                            term['files'] = [t for t in term['files'] if not templar.is_template(t)]
+                            intermediate.append(term)
+                        else:
+                            if not templar.is_template(term):
```
Comment (sivel) [inline lib/ansible/executor/task_executor.py]: These 2 lines can probably just be collapsed into an `elif`

**Comment by bmillemathias** [comment]
@sivel are you ok with the basic test I added or do you think of any corner case that should be tested while on it ?

File: lib/ansible/executor/task_executor.py (hunk 1 above)
Comment (bmillemathias) [reply lib/ansible/executor/task_executor.py]: done, I've just c/p without looking hardly on it.

**Comment by sivel** [comment]
@bmillemathias I'm not sure if we have any tests to cover the less complicated case, where you just pass files to `with_first_found` rather than a dict, containing a `files` key.  Maybe we should add a test for that variant as well

```
  with_first_found: 
         - "{{ this_var_is_empty }}"
         - "{{ role_path + '/files/foo1' }}"
```

File: test/integration/targets/lookup_first_found/tasks/main.yml [hunk 2]
```diff
@@ -71,3 +71,33 @@
...
+# https://github.com/ansible/ansible/issues/70772
+- name: test first_found with files with empty variables works (dict)
+  set_fact:
+    file_found_dict_var: "{{ item }}"
+  with_first_found:
+    - files:
+        - "{{ this_var_is_empty }}"
+        - "{{ role_path + '/files/foo1' }}"
+  vars:
+    this_var_is_empty: {}
```
Comment (sivel) [inline test/integration/targets/lookup_first_found/tasks/main.yml]: Does this replicate the behavior the user reported?  I would either not define `this_var_is_empty` at all, or do `this_var_is_empty.foo` without `foo` defined.

Comment (bmillemathias) [reply test/integration/targets/lookup_first_found/tasks/main.yml]: even with this is triggering the bug from the test I did (that why I commited first the test and then the fix to make sure it breaks then it fix) but I can do this method too.

File: lib/ansible/executor/task_executor.py [hunk 3]
```diff
@@ -19,6 +19,7 @@
 from ansible.errors import AnsibleError, AnsibleParserError, AnsibleUndefinedVariable, AnsibleConnectionFailure, AnsibleActionFail, AnsibleActionSkip
 from ansible.executor.task_result import TaskResult
 from ansible.executor.module_common import get_action_args_with_defaults
+from ansible.module_utils.common._collections_compat import Mapping
```
Comment (Akasurde) [inline lib/ansible/executor/task_executor.py:22]: Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and rebase.

File: changelogs/fragments/70819-fix-lookup-first-found-with-undefined-variable.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - fix first_found lookup with undefined variables (provided by Matt Martz)
```
Comment (Akasurde) [inline changelogs/fragments/70819-fix-lookup-first-found-with-undefined-variable.yml]:
```suggestion
  - first_found - fix undefined variables in lookup plugin (provided by Matt Martz) (https://github.com/ansible/ansible/issues/70772).
```

File: changelogs/fragments/70819-fix-lookup-first-found-with-undefined-variable.yml [hunk 5]
```diff
@@ -0,0 +1,3 @@
...
+  - add regression tests
```
Comment (Akasurde) [inline changelogs/fragments/70819-fix-lookup-first-found-with-undefined-variable.yml]:
```suggestion
```

**Pull request closed without merging**